Save the selected text size in local storage

JoelNiemela opened the pull request:
## Justification

As noted in the linked issue, the text size in the Reader's Bible does not persist after a page refresh.
This PR implemented the suggested short-term fix listed in that issue, i.e., storing the text size in local storage.

Closes #70

## What has changed

- Reads the `textSize` from local storage when the `ReadingToolbar` component is initialized.
- Updates the value in `localStorage` whenever the text size is modified using `useEffect`.

arrocke commented on src/modules/study/components/ReadingToolbar.tsx:
```diff
@@ -40,7 +40,25 @@ export default function ReadingToolbar({
   }>();
   const router = useRouter();
-  const [textSize, setTextSize] = useState(3);
+
+  // Fetch the textSize from local storage, if we're in a browser context
+  const getInitialTextSize = (): number => {
+    if (typeof window === "undefined" || typeof localStorage === "undefined") {
+      return 3;
+    }
+
+    const stored = parseInt(localStorage.getItem("ReadingToolbar-textSize") ?? "", 10);
+    return Number.isNaN(stored) ? 3 : stored;
+  };
+
+  // Persist the textSize in localStorage on update
+  const [textSize, setTextSize] = useState<number>(getInitialTextSize);
+  useEffect(() => {
+    if (typeof window !== "undefined" && typeof localStorage !== "undefined") {
+      localStorage.setItem("ReadingToolbar-textSize", textSize.toString());
+    }
+  }, [textSize]);
```
We are going to have a number of settings we want to save in local storage. Let's add a hook to `src/utils/localstorage.ts` that works like this: `const [settings, setSettings] = useLocalStorageState('reading-settings')`

It should work just like `useState` and handles all of the local storage stuff as a JSON object. That way this component can just use the hook and not worry about local storage

JoelNiemela commented:
> Thanks for the PR! I left one suggestion, and tweaked the description. We don't have user accounts for the reading view yet, so I updated the description to close the issue. We'll revisit how we implement these settings if we ever build user accounts

Ok :+1:

One thing I noticed when addressing the review comment above: it seems like we use server-side rendering for the read page (e.g., `/en/read/eng/01001`), and this seems to occasionally mess with the text size occasionally. As local storage is obviously not available on the server, the defaults are used in that case, which means that the text is initially rendered in the default text size. Usually an update is triggered right after the page loads, causing the text to be re-rendered with the text size saved in local storage, but occasionally this doesn't seem to happen.

I'm unsure what the correct fix to this is (and it might not be a significant problem — my dev server runs much slower than the production server does, so it might be caused by that?), but I wanted to at least point that out.

JoelNiemela commented on src/utils/localstorage.ts:
```diff
@@ -0,0 +1,37 @@
+    if (typeof window !== "undefined" && typeof localStorage !== "undefined") {
+      localStorage.setItem(item, JSON.stringify(settings));
+    }
+  }, [settings]);
```
Fixed.

JoelNiemela commented on src/utils/localstorage.ts:
```diff
@@ -0,0 +1,37 @@
+    const json: T = {} as T;
+    for (const key in defaultValue) {
+      json[key] = key in parsed ? parsed[key] : defaultValue[key];
+    }
```
Ah, I didn't think of simply spreading the two objects like that.

Fixed. (With the difference that I removed the last `if (parsed)` check, as it's not required (you can safely spread any value, including `null`/`undefined`, apparently), and I added a `?? ''` after `localStorage.getItem` to make TypeScript happy).

arrocke commented on src/utils/localstorage.ts:
```diff
@@ -0,0 +1,30 @@
+
+export default function useLocalStorageState<T extends Record<string, any>>(item: string, defaultValue: T): [T, Dispatch<SetStateAction<T>>] {
+  // Fetch the settings from local storage if we're in a browser context
+  const getInitialState = (): T => {
```
Ok, I think I understand the issue you are seeing. We need to do a couple things here:
1. We need to mark this with `"use client";` since this is a hook that can only be executed on the client side.
2. We need to initialize the state from local storage using `useLayoutEffect`. This ensures that the initialization code runs on the client side since it is bypassed on the server side.
3. If we make that change, then I don't think we need to check that localStorage exists because the effect will only run in the browser

The hook will look something like this:
```typescript
function useLocalStorageState(item, defaultValue) {
  const [state, setState] = useState(defaultValue)

  useEffect(() => {
    // Initialize state from localstorage
  }, [item])

  useEffect(() => {
    // Update local storage
  }, [item, state])

  return [state, setState]
}
